src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
+// see above To summarized:
+//  IF I understand this correctly:
+//    for each database id on this shared: (this is only 0)
+//      lock each key - don't
+//      check memory utilization for the address  of the key/value -> only for the value
+//      move the key/value to new location  -> just do allocation (for the value), move data to new
+//      memory location and free old one.
+//   The for each is running by the task itself, and the cursor is the for index - again for now
+//   only index 0!
+
+auto ReallocateValueImpl(PrimeValue& current_value) -> bool {
+  // we know that this is a string, we checked that before
+  auto val = current_value.ToString();
+  // we can now reset the old object as we are going to create a new one
+  current_value.Reset();
+  // This should allocate a new value
+  current_value.SetString(val);
+  return current_value.MallocUsed() > 0;
+}
+
+auto ReallocateValue(PrimeIterator it, void* from, const MemoryDefragParams& params, DbIndex index)
+    -> bool {
+  // allocate a new memory for the value, and then free the memory that used for the old one
+  CHECK(it->second.ObjType() == OBJ_STRING);  // only supporting for string right now!
+  if (it->second.IsExternal()) {
+    VLOG(1) << std::hex << from << ": the object that we are dealing with is external for type "
+            << std::dec << OpTypeToString(params.object_type)
+            << " or was not allcoated on the heap";
+    return false;
+  }
+
+  params.slice->PreUpdate(index, it);
+  auto key = it->first.ToString();
+  auto state = ReallocateValueImpl(it->second);
+  params.slice->PostUpdate(index, it, key);
+  return state;
+}
+
+auto MemoryReallocate(PrimeIterator it, const MemoryDefragParams& params, DbIndex index)
+    -> MemDefragState {
+  if (!IsValid(it)) {
+    return MemDefragState::REQUIRED;
+  }
+  if (MatchType(it->second.ObjType(), params.object_type)) {
+    auto ptr = it->second.RObjPtr();
+    return ReallocateValue(it, ptr, params, index) ? MemDefragState::REQUIRED
+                                                   : MemDefragState::NOT_REQUIRED;
+  }
+  return MemDefragState::NOT_REQUIRED;
+}
+
+auto DefragMemory(const MemoryDefragParams& params, DbIndex index)
+    -> std::tuple<MemDefragState, uint64_t> {
+  constexpr size_t CURSOR_COUNT_RELATION = 50;
+
+  auto shard_id = params.slice->shard_id();
+  DCHECK(params.slice->IsDbValid(index));
+
+  unsigned cnt = 0;
+  std::size_t max = params.slice->DbSize(index) / CURSOR_COUNT_RELATION;
+  VLOG(1) << "PrimeTable shared id/index " << params.slice->shard_id() << "/" << index << " has "
+          << params.slice->DbSize(index) << " elements in it";
+
+  PrimeTable::Cursor cur = params.cursor;
+  auto [prime_table, expire_table] = params.slice->GetTables(index);
+  MemDefragState state = max > 0 ? MemDefragState::REQUIRED : MemDefragState::NOT_REQUIRED;
+
+  while (cnt < max && state != MemDefragState::NOT_REQUIRED) {
```

Comment:
using the cursor to iterate on the data here, just like in scan

Reply:
why do you use `state` here. Traverse has a well defined API where it goes over defined iterators. what added value `state` gives you *here*?

Reply:
removed